src/rules/at-root-no-redundant-selector/index.js

```diff
@@ -0,0 +1,74 @@
+"use strict";
+
+const { utils } = require("stylelint");
+const namespace = require("../../utils/namespace");
+const ruleUrl = require("../../utils/ruleUrl");
+
+const ruleName = namespace("at-root-no-redundant-selector");
+
+function isInterpolated(string) {
+  const interpolationRegex = /#{(.*?)}/g;
+  return interpolationRegex.test(string);
+}
+
+function isAtRootRuleNested(node) {
+  // Recursive method to check if any descendant is an @at-root rule.
+  let isAtRootNested = false;
+  if (node.nodes !== undefined) {
+    for (const i in node.nodes) {
+      isAtRootNested = isAtRootNested || isAtRootRuleNested(node.nodes[i]);
+      if (isAtRootNested) break;
+    }
+  }
+  return isAtRootNested || (node.type === "atrule" && node.name === "at-root");
+}
+
+const messages = utils.ruleMessages(ruleName, {
+  rejected: "Unnecessary @at-root rule"
+});
+
+const meta = {
+  url: ruleUrl(ruleName)
+};
+
+function rule(actual) {
+  return (root, result) => {
+    const validOptions = utils.validateOptions(result, ruleName, {
+      actual
+    });
+
+    if (!validOptions) {
+      return;
+    }
+
+    root.walkAtRules(decl => {
+      if (decl.type !== "atrule") return;
+      if (
+        decl.name === "at-root" &&
+        (decl.parent.type === "root" ||
+          (decl.params.includes("&") && !isInterpolated(decl.params)))
+      ) {
+        utils.report({
+          message: messages.rejected,
+          node: decl,
+          result,
+          ruleName
+        });
+      }
+      if (decl.name === "keyframes" && isAtRootRuleNested(decl)) {
+        utils.report({
+          message: messages.rejected,
+          node: decl,
```

Comment:
You don't want to report the error on the `@keyframes`, instead you want to report it on the `@at-rule` node.

You can implement the suggestion from above with `findAtRoot`, and report it there, e.g.:
   ```js
   if (node.name === "keyframes") {
     const atRoot = findAtRoot(node);
     if (atRoot) {
       utils.report({
         message: messages.rejected,
         node: atRoot,
         result,
         ruleName
       });
     }
   }
   ```